bugsnag/bugsnag-android
# Read projectPackages array when serialising error reports

## Goal

It is currently possible to call `Bugsnag.setProjectPackages` after bugsnag has been initialsed, although for nearly all conceivable usecases this value won't change in the lifetime of an application. For example:

```
Bugsnag.init(this) // uses the default project packages, and starts flushing error reports
Bugsnag.setProjectPackages("com.myexample") // sets value AFTER report is sent
```

The `Bugsnag` and `Client` accessors have been deprecated in favour of setting this value on the `Configuration` object, and passing that in as a constructor parameter to the `Client`. Additionally, the `ErrorReader` class has been altered so that the correct `projectPackages` value is now used.

## Changeset

- Deprecated `projectPackages` accessors on `Bugsnag` and `Client` in favour of `Configuration`
- Moved `inProject()` to `Stacktrace` as only used there
- Added `projectPackages` field to `Stacktrace` that is used for in-project calculations
- Altered test code to reflect new call signatures
- Made `ErrorReader` serialisation methods generic with casts so caller can use typing properly

## Tests

Manually tested

Followed the steps in #418 to confirm that the `projectPackages` is now correctly set when manually inspecting the payload. Also relied on existing unit tests.
